### PR TITLE
JSUI-2798 Improved accessibility of `Quickview`

### PR DESCRIPTION
**Pull request by btaillon**
https://coveord.atlassian.net/browse/JSUI-2798

1. Added label to close button
2. Added tab index to close button for keyboard navigation
3. Locked keyboard focus within `QuickView`'s modal while it's open.
4. Added accessibility tests for QuickView

This was tested with Android's TalkBack and doesn't seem to work. Most tutorials and articles online, even when searching specifically for accessible modals, seem to ignore accessibility altogether by detecting the `tab` key. However, some tutorials using ally.js seem to detect focus/blur for the desktop and add `aria-hidden` for mobile accessibility tools, which seems to be a less radical solution than setting `tabindex="-1"` to every focusable element. There is no guarantee that this works with iOS's VoiceOver either, but setting `aria-hidden` and `aria-modal` seems to be the recommended method by W3 (https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/dialog.html).

[Deploy](https://dashboard.heroku.com/pipelines/a3535101-5bbf-4a5b-a909-47fcf8c9f149)

### CODE REVIEW

File: lib/modal-box/index.d.ts [hunk 1]
```diff
@@ -1,96 +1,97 @@
-declare module Coveo.ModalBox {
+declare namespace Coveo.ModalBox {
```
Comment (btaillon) [inline lib/modal-box/index.d.ts]: I tried to revert this change along with the tabbing, but it looks like whenever I commit my changes, the file gets re-formatted like this. I'll write comments under the lines I changed in this file.

File: lib/modal-box/index.d.ts [hunk 2]
```diff
@@ -1,96 +1,97 @@
...
+    /**
+     * Specify a validation function, which receives the button that was pressed.<br/>
+     * If the validation function return true, the modal box closes, otherwise it stays open
+     * @param button
+     */
+    validation?: (button: BUTTON) => boolean;
+    /**
+     * Specify the title of the modal box
+     */
+    title?: string | HTMLElement;
```
Comment (btaillon) [inline lib/modal-box/index.d.ts]: This is one of the changes I made. I changed this type from `string` to `string | HTMLElement` since some old code was already setting an `HTMLElement` on this property.

File: lib/modal-box/index.d.ts [hunk 3]
```diff
@@ -1,96 +1,97 @@
...
+    /**
+     * Specify the content that you wish to put inside the modal box
+     */
+    body?: HTMLElement;
+    sizeMod?: 'big';
```
Comment (btaillon) [inline lib/modal-box/index.d.ts]: This is the other change I made: I added this missing property to avoid getting errors. I don't know what other values it can get than `big`, so I only set it to the type `'big'` for now.

File: src/ui/Quickview/Quickview.ts [hunk 4]
```diff
@@ -453,10 +464,64 @@ export class Quickview extends Component {
...
+        return;
+      }
+    });
+    $$(document.body).on('keydown', (e: KeyboardEvent) => {
+      if (!this.modalbox) {
+        return;
+      }
+      if (e.keyCode !== KEYBOARD.TAB) {
+        return;
+      }
```
Comment (btaillon) [inline src/ui/Quickview/Quickview.ts]: I'm worried this might not work on some screen readers. Is there some tools I should test this with?

File: accessibilityTest/AccessibilityQuickview.ts [hunk 5]
```diff
@@ -1,33 +1,101 @@
...
+import { afterQuerySuccess, getRoot, testResultElement } from './Testing';
+
+function nodeListToArray<T extends Node>(nodeList: NodeListOf<T>) {
+  const result: T[] = [];
+  for (let i = 0; i < nodeList.length; i++) {
+    result.push(nodeList.item(i));
+  }
+  return result;
+}
+
```
Comment (olamothe) [inline accessibilityTest/AccessibilityQuickview.ts]: Could simply use `Dom.nodeListToArray` instead of rewriting it ?

File: accessibilityTest/AccessibilityQuickview.ts [hunk 6]
```diff
@@ -1,33 +1,101 @@
...
-      await afterDelay(500);
+      await openQuickview();
       const axeResults = await axe.run(getRoot());
       expect(axeResults).toBeAccessible();
       done();
     });
+
+    it('should have a close button when opened', async done => {
+      await openQuickview();
+      expect(document.querySelector('.coveo-small-close')).not.toBeNull();
```
Comment (olamothe) [inline accessibilityTest/AccessibilityQuickview.ts]: I would create a function for `document.querySelector('.coveo-small-close')` -> `getQuickviewCloseButton()`

File: accessibilityTest/AccessibilityQuickview.ts [hunk 7]
```diff
@@ -1,33 +1,101 @@
...
+      const container: HTMLElement = document.querySelector('.coveo-modal-container');
+      const initialFocus = document.activeElement;
+      const maxAttempts = 200;
+      for (let attempt = 0; attempt < maxAttempts; attempt++) {
+        simulateTabKey();
+        if (document.activeElement === initialFocus || !container.contains(document.activeElement)) {
+          break;
+        }
+      }
+      expect(document.activeElement.outerHTML).toBe(initialFocus.outerHTML);
```
Comment (olamothe) [inline accessibilityTest/AccessibilityQuickview.ts]: Can you instead test directly for `expect(document.activeElement).toBe(initialFocus)` ? Instead of outerHTML ?

File: src/ui/Quickview/Quickview.ts (hunk 4 above)
Comment (olamothe) [reply src/ui/Quickview/Quickview.ts]: You can use osx voiceover.

File: src/ui/Quickview/Quickview.ts [hunk 8]
```diff
@@ -453,10 +464,64 @@ export class Quickview extends Component {
...
+      this.makeModalboxAccessible(this.modalbox.modalBox);
       return computedModalBoxContent;
     });
   }
 
+  private getFocusableElements(container: HTMLElement) {
+    return _.sortBy(container.querySelectorAll<HTMLElement>('[tabindex]'), element => element.tabIndex);
+  }
+
+  private createFocusTrap(element: HTMLElement) {
```
Comment (olamothe) [inline src/ui/Quickview/Quickview.ts]: Was creating such a focusTrap the recommended approach on dequeue university ?

Also, this whole function is pretty hard to follow/read.

Id split/break the logic up in different method. 
I could even see all of this moved to `FocusTrap` module/class maybe ?

Which accept a `container` or root element inside which the trap should be created, and the tab cycling logic would be implemented.

File: src/ui/Quickview/Quickview.ts [hunk 9]
```diff
@@ -453,10 +464,64 @@ export class Quickview extends Component {
...
+    document.addEventListener('focusin', e => {
+      if (!this.modalbox) {
+        return;
+      }
+      if (!element.contains(e.target as HTMLElement)) {
+        const focusableElements = this.getFocusableElements(element);
+        focusableElements[0].focus();
+        e.preventDefault();
```
Comment (olamothe) [inline src/ui/Quickview/Quickview.ts]: Why is the `preventDefault` needed here ?

File: src/ui/Quickview/Quickview.ts [hunk 10]
```diff
@@ -453,10 +464,64 @@ export class Quickview extends Component {
...
+    return _.sortBy(container.querySelectorAll<HTMLElement>('[tabindex]'), element => element.tabIndex);
+  }
+
+  private createFocusTrap(element: HTMLElement) {
+    document.addEventListener('focusin', e => {
+      if (!this.modalbox) {
+        return;
+      }
+      if (!element.contains(e.target as HTMLElement)) {
+        const focusableElements = this.getFocusableElements(element);
```
Comment (olamothe) [inline src/ui/Quickview/Quickview.ts]: What if there's no focusable elements returned ?

File: src/ui/Quickview/Quickview.ts [hunk 11]
```diff
@@ -453,10 +464,64 @@ export class Quickview extends Component {
...
+      const isGoingForward = !e.shiftKey;
+      if (isGoingForward) {
+        if (activeElement === _.last(focusableElements)) {
+          focusableElements[0].focus();
+          e.preventDefault();
+        }
+      } else {
+        if (activeElement === focusableElements[0]) {
+          _.last(focusableElements).focus();
+          e.preventDefault();
```
Comment (olamothe) [inline src/ui/Quickview/Quickview.ts]: Need to make sure it won't break if there's somehow no focusable element.

File: src/ui/Quickview/Quickview.ts [hunk 12]
```diff
@@ -453,10 +464,64 @@ export class Quickview extends Component {
...
       return computedModalBoxContent;
     });
   }
 
+  private getFocusableElements(container: HTMLElement) {
+    return _.sortBy(container.querySelectorAll<HTMLElement>('[tabindex]'), element => element.tabIndex);
+  }
+
+  private createFocusTrap(element: HTMLElement) {
+    document.addEventListener('focusin', e => {
```
Comment (olamothe) [inline src/ui/Quickview/Quickview.ts]: This function will get created everytime a new quickview is opened, but it never gets cleaned up when the modal box closes. That's problematic.

File: src/ui/Quickview/Quickview.ts [hunk 13]
```diff
@@ -453,10 +464,64 @@ export class Quickview extends Component {
...
+        return;
+      }
+      if (!element.contains(e.target as HTMLElement)) {
+        const focusableElements = this.getFocusableElements(element);
+        focusableElements[0].focus();
+        e.preventDefault();
+        return;
+      }
+    });
+    $$(document.body).on('keydown', (e: KeyboardEvent) => {
```
Comment (olamothe) [inline src/ui/Quickview/Quickview.ts]: Same thing about the function being created everytime a new quickview is created but never cleaned up.

File: unitTests/Simulate.ts [hunk 14]
```diff
@@ -221,6 +221,31 @@ export class Simulate {
     return modalBox;
   }
 
+  static fullModalBoxModule(): ModalBox {
```
Comment (olamothe) [inline unitTests/Simulate.ts]: Why did you create a new function for this instead of modifying the existing `modalBoxModule` ? It seems very hackish ;)

File: accessibilityTest/AccessibilityQuickview.ts (hunk 5 above)
Comment (btaillon) [reply accessibilityTest/AccessibilityQuickview.ts]: Woops, I wasn't aware this function existed. I'll use it instead.

File: accessibilityTest/AccessibilityQuickview.ts (hunk 6 above)
Comment (btaillon) [reply accessibilityTest/AccessibilityQuickview.ts]: Good idea, fixing that.

File: accessibilityTest/AccessibilityQuickview.ts (hunk 7 above)
Comment (btaillon) [reply accessibilityTest/AccessibilityQuickview.ts]: Woops, I used this because jasmine gave ambiguous error messages otherwise. I forgot to fix this, I'll remove `.outerHTML`.

File: src/ui/Quickview/Quickview.ts (hunk 8 above)
Comment (btaillon) [reply src/ui/Quickview/Quickview.ts]: I couldn't find any recommendation from Deque other than to make sure elements outside of a modal aren't focusable.

> If a screen reader user navigates by elements such as headings, landmarks, links, form elements, etc., the only elements available when the dialog is open are the elements in the dialog itself. There must not be a way to access anything outside of the dialog with screen reader keyboard shortcuts.

As for the code being hard to read, I agree that it could be broken up into a separate class and I'll do that.

File: src/ui/Quickview/Quickview.ts (hunk 9 above)
Comment (btaillon) [reply src/ui/Quickview/Quickview.ts]: `preventDefault` is needed here to prevent the browser from changing the currently focused element to one that's outside of the `Quickview`.

File: src/ui/Quickview/Quickview.ts (hunk 10 above)
Comment (btaillon) [reply src/ui/Quickview/Quickview.ts]: Since the focus trap is currently only implemented for the `Quickview`, it always has at-least one focusable element (the close button). However, I agree that it would be a good idea to add a check here. I'll add one.

File: src/ui/Quickview/Quickview.ts (hunk 11 above)
Comment (btaillon) [reply src/ui/Quickview/Quickview.ts]: I'll fix that.

File: src/ui/Quickview/Quickview.ts (hunk 12 above)
Comment (btaillon) [reply src/ui/Quickview/Quickview.ts]: Woops, I didn't think about the fact that those are now on `document` and need to be cleaned up. I'm going to fix this.

File: src/ui/Quickview/Quickview.ts (hunk 13 above)
Comment (btaillon) [reply src/ui/Quickview/Quickview.ts]: I'll clean up function too.

File: unitTests/Simulate.ts (hunk 14 above)
Comment (btaillon) [reply unitTests/Simulate.ts]: I did that because the context I needed to test `Quickview` was too different from the already existing one and I didn't want to risk breaking other accessibility tests. I'm going to test this one with every accessibility test and see if it works.

**Comment by btaillon** [comment]
I added the `FocusTrap` component, which should be easier to test and re-use in other components if needed. I also added a few unit tests to go along with it and avoid regressions.

File: src/ui/FocusTrap/FocusTrap.ts [hunk 15]
```diff
@@ -0,0 +1,112 @@
+import _ = require('underscore');
```
Comment (ThibodeauJF) [inline src/ui/FocusTrap/FocusTrap.ts]: Import only the few methods you need for underscore

File: src/ui/FocusTrap/FocusTrap.ts [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+  private focusSibling(element: HTMLElement, previous = false) {
+    const sibling = this.getSibling(element, previous);
+    if (sibling) {
+      sibling.focus();
+    }
+  }
+
+  private focusFirstElement() {
+    const elements = this.focusableElements;
+    if (elements.length > 0) {
```
Comment (ThibodeauJF) [inline src/ui/FocusTrap/FocusTrap.ts]: could only use `if (elements.length)`

Comment (btaillon) [reply src/ui/FocusTrap/FocusTrap.ts]: Yeah that's what I initially wrote but I wasn't sure which option was preferable. I'll change it to `elements.length`.

File: src/ui/FocusTrap/FocusTrap.ts (hunk 15 above)
Comment (btaillon) [reply src/ui/FocusTrap/FocusTrap.ts]: Woops, fixing that.
